Distant light source plugin

leroyvn opened the pull request:
This is a tentative implementation of a distant light source plugin. A test scene file is included in the `resources/test_scenes` directory and should give something like this:
![test_distant](https://user-images.githubusercontent.com/34740232/78159987-9d5d7500-7443-11ea-9a23-f2b508c1cf83.png)

Speierers commented on src/emitters/distant.cpp:
```diff
@@ -0,0 +1,157 @@
+        Vector3f v = m_world_transform->eval(si.time, active)
+                         .inverse()
+                         .transform_affine(-si.wi);
+        return (allclose(v, Vector3f{ 0.f, 0.f, 1.f }))
+                   ? unpolarized<Spectrum>(m_irradiance->eval(si, active))
+                   : unpolarized<Spectrum>(0.f);
```
Here you should use a `select` statement as `v` might be a array of vectors (e.g. GPU variant).

Speierers replied:
Also, this light having a `DeltaDirection`, should `eval` always return `0.f`?

leroyvn replied:
Regarding the use of `select`, I'm still missing an equivalent to `allclose` which would only operate on each packet channel. This is not working as intended:
```suggestion
        return select(allclose(v, Vector3f{ 0.f, 0.f, 1.f }),
                      unpolarized<Spectrum>(m_irradiance->eval(si, active)),
                      unpolarized<Spectrum>(0.f));
```

As for the return value of `eval`, came up with this behaviour by reading the Mitsuba 1 implementation; but I honestly don't know what all the consequences could be, maybe I could miss something?

Speierers replied:
Mitsuba 2 doesn't have this notion of `measure`, I would say we are always in `ESolidAngle` therefore `eval` and `pdf` should always return `0.f`

leroyvn replied:
How about that? I basically did what is in `allclose`, without the nested horizontal operation.

```suggestion
        constexpr float rtol = 1e-5f;
        constexpr float atol = 1e-8f;

        auto mask = all(abs(ScalarVector3f{ 0.f, 0.f, 1.f } - v) <=
                        abs(v) * rtol + atol);

        return select(mask, 
                      unpolarized<Spectrum>(m_radiance->eval(si, active)),
                      unpolarized<Spectrum>(0.f));
```

Speierers replied:
I would still argue that this is wrong as the probability of sampling the peak of a delta distribution is zero. You can have a look at `dielectric.cpp` which is a `Delta` BSDF (same idea). There `eval` and `pdf` always return `0.f`, so it should be the same here.

leroyvn replied:
Ok, I fixed that.

Speierers commented on src/emitters/distant.cpp:
```diff
@@ -0,0 +1,157 @@
+                                          const Point2f & /*sample3*/,
+                                          Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointSampleRay, active);
+        NotImplementedError("sample_ray");
```
This method will be useful in the future (e.g. for a bidirectional path tracer). It would be great to have it implemented here.

leroyvn replied:
I had an implementation for it in the beginning, I'll put it back here.

Speierers commented on src/emitters/distant.cpp:
```diff
@@ -0,0 +1,157 @@
+        si.wavelengths          = it.wavelengths;
+
+        return std::make_pair(
+            ds, unpolarized<Spectrum>(m_irradiance->eval(si, active)) / ds.pdf);
```
Not necessary to divide by pdf here as it is alway `1.f`

leroyvn replied:
Noted.

Speierers commented on src/emitters/distant.cpp:
```diff
@@ -0,0 +1,157 @@
+            ds, unpolarized<Spectrum>(m_irradiance->eval(si, active)) / ds.pdf);
+    }
+
+    Float pdf_direction(const Interaction3f &, const DirectionSample3f &ds,
+                        Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointEvaluate, active);
+
+        Vector3f d = m_world_transform->eval(ds.time, active)
+                         .transform_affine(Vector3f{ 0.f, 0.f, 1.f });
+        return allclose(ds.d, -d) ? 1.0f : 0.0f;
+    }
```
Should always return `0.f` as it is a `DeltaDirection` light.

leroyvn replied:
I also got this from the Mitsuba 1 implementation; should I change it?

leroyvn replied:
It's fixed.

Speierers commented on src/emitters/distant.cpp:
```diff
@@ -0,0 +1,157 @@
+        }
+
+        m_flags      = EmitterFlags::Infinite | EmitterFlags::DeltaDirection;
+        m_irradiance = props.texture<Texture>("irradiance", Texture::D65(1.f));
```
Maybe we could change `irradiance` to `radiance` throughout this whole file for consistency with `area.cpp` and other light source?

leroyvn replied:
You're right, we're actually returning a radiance (not divided by `$\pi$`). I'll fix that.

Speierers commented on src/emitters/distant.cpp:
```diff
@@ -0,0 +1,157 @@
+#include <mitsuba/core/bsphere.h>
+#include <mitsuba/core/plugin.h>
+#include <mitsuba/core/properties.h>
+#include <mitsuba/core/warp.h>
```
Remove unnecessary includes here

leroyvn replied:
Noted.

Speierers commented on src/emitters/tests/test_distant.py:
```diff
@@ -0,0 +1,119 @@
+    # Check correctness of the eval() method
+    from mitsuba.render import SurfaceInteraction3f
+
+    direction = np.array([0, 0, -1])
```
please use `mitsuba.core.Vector3f` and remove `import numpy as np` in this test file

Speierers commented on src/emitters/tests/test_distant.py:
```diff
@@ -0,0 +1,119 @@
+
+    from mitsuba.render import SurfaceInteraction3f
+
+    direction = np.array([0, 0, -1])
```
Maybe do a `for` loop over a couple directions here?

leroyvn replied:
I parametrised the tests with a set of arbitrary directions.

Speierers commented on src/emitters/distant.cpp:
```diff
@@ -0,0 +1,157 @@
+            m_world_transform =
+                new AnimatedTransform(ScalarTransform4f::look_at(
+                    ScalarPoint3f(0.0f), ScalarPoint3f(direction), up));
+        }
```
should there be an `else` statement for when only `to_world` is defined?

leroyvn replied:
I guess so, although I actually don't really see a use case for that. I also realise that there is not default, I'll add one.

Speierers commented on resources/test_scenes/test_distant.xml:
```diff
@@ -0,0 +1,44 @@
+<scene version="2.0.0">
```
This file should be moved to `resources/data/tests/scenes/somedirectory` so it will be picked by the rendergin test suite as described [here](https://mitsuba2.readthedocs.io/en/latest/src/developer_guide/testing.html#rendering-test-suite-and-z-test)

leroyvn replied:
Should I fork the data repo as well and create a PR there?

Speierers replied:
Indeed, and have a commit here that updates the submodule.

leroyvn commented:
@Speierers I pushed a couple of commits which fix some of the points you raised. I also renamed the plugin `directional` to remain consistent with Mitsuba 1 and improve scene file compatibility.

The following points remain open:

- I still have to fix the `eval` and `pdf_direction` methods (I missed your reply).
- I'm now realising that the Mitsuba 1 plugin had `irradiance` as its parameter name; I don't know what to do (consistency with `area` vs scene file comptability).
- I couldn't find out how to instantiate an identity `Transform` and I'm not happy with the workaround I used (scale with factor 1).
- The `sample_ray` method still needs care, I'll need to discuss a bit more.

Speierers commented on src/emitters/directional.cpp:
```diff
@@ -59,7 +60,8 @@ class DirectionalEmitter final : public Emitter<Float, Spectrum> {
             m_world_transform = props.animated_transform("to_world");
         } else {
-            m_world_transform = new AnimatedTransform(ScalarTransform4f::scale(1.));
+            m_world_transform =
```
I just realized that the `Endpoint` constructor already initializes the `m_wolrd_transform` transform if provided. So you original code was correct.

Speierers commented:
Great progress!

> * I still have to fix the `eval` and `pdf_direction` methods (I missed your reply).
I just replied above.

> * I'm now realising that the Mitsuba 1 plugin had `irradiance` as its parameter name; I don't know what to do (consistency with `area` vs scene file comptability).
Okay then let's switch back to `irradiance` to be consistent with Mitsuba 1 (sorry for that)

> * I couldn't find out how to instantiate an identity `Transform` and I'm not happy with the workaround I used (scale with factor 1).
See comment in the code.

> * The `sample_ray` method still needs care, I'll need to discuss a bit more.
See comments in the code

Speierers commented on src/emitters/directional.cpp:
```diff
@@ -89,13 +91,33 @@ class DirectionalEmitter final : public Emitter<Float, Spectrum> {
+
+        // 2. Sample spatial component
+        // TODO: sample a point on such that cast ray will hit the scene
+        Vector3f v0 = warp::square_to_uniform_sphere(sample2);
```
This could should be similar to the Mitsuba 1 implementation (sample on disk and transform)

leroyvn replied:
See latest commits for an implementation.

leroyvn commented:
Thanks a lot for your feedback and help. I fixed the `eval` and `pdf_position` methods, rolled back the parameter name change and the default transform argument, and proposed an implementation for the `sample_ray` method. For the latter, I essentially reproduced what's in Mitsuba 1.

I wrote the tests for the scalar RGB variant but I also tried in packet mode (just one line needs to be modified for that purpose). I'm not sure how I should check wavelengths.

On a separate note, I wonder: shouldn't be spectral sampling handled by the attached spectrum (`m_irradiance`) rather than the `sample_wavelength` template function? 

In general, I think it would be interesting to allow for more flexibility with respect to spectral sampling (I'm using Mitsuba for applications other than rendering and spectral features are critical for us). I can open a discussion on that topic if you think it's appropriate.

Speierers commented:
> In general, I think it would be interesting to allow for more flexibility with respect to spectral sampling (I'm using Mitsuba for applications other than rendering and spectral features are critical for us). I can open a discussion on that topic if you think it's appropriate.

Not sure how to answer this. Let's open an issue for that so we can discuss this futher there!

Speierers commented:
Everything looks good to me! 

Could you maybe squash this PR into a single commit and add the scene to the rendering test suite?

Once this is done we should be able to rebase and merge this PR to master 🚀 

Thanks a lot for your contribution!

leroyvn commented:
I squashed (and fixed a missing directory creation in `test_renders.py`) but haven't updated yet the data submodule. I'll wait until my PR there is accepted, then I'll update the submodule by amending this commit and we'll be ready to merge.

Speierers commented:
Great! Just reduce the resolution of the test render and then we will be able to merge !

Speierers commented:
I just merged the PR in mitsuba-data. Go ahead add the commit to update the submodule

leroyvn commented:
Done

Speierers commented:
Thanks again!

leroyvn commented:
Thank you as well for the review and assistance!
